Memory mapped file support for reading Ogawa files

ianw-foundry opened the pull request:
This change adds memory mapped files support for POSIX (tested on CentOS7/Linux) and Windows systems. It is based on changes I originally made to Alembic 1.6.1, as used by Foundry's Katana product.

Performance tests using AbcWalk on a small collection of Ogawa format files show a modest improvement in wall-clock runtime for Linux and a larger improvement on Windows. My fairly simple test case show about a reduction from ~35s to ~27s on Linux, and from ~170s to ~50s on Windows.

There's a few things that I'd be happy to update! These are:

1. Passing an options object for selecting memory mapped or stream based file access. This uses dynamic_cast from a base class with a dummy virtual function I've put in Util/Foundation.h. It would allow different options to be passed to different layers without them 'knowing' of each other's options... but it's a bit heavy-weight for just one option. Is there an alternative that would fit better with Alembic?

2. The default option is currently set to use memory mapped files. This is merely convenient for my testing, and not an opinion on suggested policy.

3. I've disabled the test `testWriteWhileRead`. This test overwrites an already opened file, and checks the behaviour of further reads. Detecting this case and having it fail with a read-error with a memory mapped file is a little tricky to do without impacting performance. There are a few alternatives, including preventing the file being opened for writing, though this would be a change of this tests behaviour and may be difficult to consistently do between OSs. Is there an alternative behaviour which would be acceptable?

4. What documentation would be suitable? Where should it go?

Cheers!
Ian W

lamiller0 commented:
1.  Yeah, I agree having an options struct like that is a bit heavy and intrusive.   I would probably just have a boolean and set it up along with AbcCoreFactory/IFactory much in the same way it supports setOgawaNumStreams.  I'm not currently aware of a good reason to have different layers use different options

Is there a platform that one would want to build Alembic on that doesn't support memory mapped reads?  It would help clean up some layered preprocessor defines in IStreams.cpp

For 3.  we can just enable the test but test it for traditional reading only.

lamiller0 commented on CMakeLists.txt:
```diff
@@ -137,7 +137,7 @@ ENDIF()
 IF ("${CMAKE_BUILD_TYPE}" MATCHES "Debug")
     ADD_DEFINITIONS(-DDEBUG=1 -UNDEBUG)
     IF (NOT WINDOWS)
-        ADD_DEFINITIONS(-Wall -Werror -Wextra -Wno-unused-parameter)
```
Why was this pulled out?

ianw-foundry replied:
Ah, sorry about that.

When building Alembic against some older versions of HDF5, debug builds would fail due to warnings in HDF5Public.h.

Removing it was a convenience on my local branch that I should have excluded from the branch I pushed with the other changes. 

I'll put it back in shortly!

ianw-foundry commented:
> I would probably just have a boolean and set it up along with AbcCoreFactory/IFactory much in the same way it supports setOgawaNumStreams.

Options object has been removed, and a simple boolean is passed from IFactory to the Ogawa IStreams constructor. In the IFactory, I've included an enum rather than just used a naked bool so it's a little more clear what the option are between.

> Is there a platform that one would want to build Alembic on that doesn't support memory mapped reads? It would help clean up some layered preprocessor defines in IStreams.cpp

I agree it's unlikely Alembic needs to be built on other platforms. I've simplified some of the conditional compilation a little. It should be fine on Windows, OSX, Linux and other unix-like's that has POSIX mmap. I've only directly built and tested on Linux and Windows.

> For 3. (the testWriteWhileRead test) we can just enable the test but test it for traditional reading only.

I've updated the ArrayPropertyTests to run for both file streams and mmap. The testWriteWhileRead test is only ran when using file streams.

Do we want to make a similar change to the other AbcCoreOgawa/Test, so that the test cases are ran for both?

lamiller0 commented:
Yes please, it should test both in AbcCoreOgawa/Test.

meshula commented:
> Is there a platform that one would want to build Alembic on that doesn't support memory mapped reads?

memmap is awesome on iOS and macOS, so support for memmap would be great for the OS level integrations of Alembic.

ianw-foundry commented:
Cheers @lamiller0 

Do you have any performance results you may have ran with this change? I've ran it against some of the assets that we have here at the Foundry for QA purposes, and seen good results. You've almost certainly got a better set of test files available, though. It would be good to verify that the speedups we see on our test files also applies across a wider test set.

Thanks again!

lamiller0 commented:
I've tested it on Linux on a couple of larger static environments (so hundreds of Alembic files) in Katana, and was measuring about a 10% speedup in the Alembic_In op.

I've also tested it on some bigger animated assets, not in Katana, but in a modified AbcWalk and have seen a 10% to 25% speedup.

Unfortunately I don't have access to our assets on a Windows machine, so I can't verify it there.

ianw-foundry commented:
Thanks Lucas. Good to have confirmation! :)
